### PR TITLE
PAYARA-3885 Implement Parameterized dynamically resolved role names

### PR DESCRIPTION
**Pull request by jGauravGupta**
Implemented Parameterized dynamically resolved role names feature by using existing annotation @RolesPermitted: e.g 

```
@RolesPermitted("#{'CustomerCare_'.concat(customer.name)}")
public void updateCustomer(@Named("customer") Customer customer){
}
```

In the above example, `Named` annotation is used to define the parameter name in dynamic role expression. 

Multiple parameters can be used in el expression:  e.g 
```
@RolesPermitted("#{self.findRole(customer, product)}")
public void updateCustomer(@Named("customer") Customer customer, @Named("product") Product product){
}
```

If the parameter name not defined using `Named` annotation and method parameter count is 1 then default `param` name is used, e.g 
```
@RolesPermitted("#{'CustomerCare_'.concat(param.name)}")
public void updateCustomer(Customer customer){
}
```

Testcases: https://github.com/payara/payara-samples/pull/24

### CODE REVIEW

**Comment by jGauravGupta** [comment]
jenkins test please

**Comment by jGauravGupta** [comment]
jenkins test please

**Comment by jGauravGupta** [comment]
jenkins test please

**Comment by jGauravGupta** [comment]
jenkins test please

File: appserver/payara-appserver-modules/cdi-auth-roles/src/main/java/fish/payara/appserver/cdi/auth/roles/RolesPermittedInterceptor.java [hunk 1]
```diff
@@ -66,47 +66,69 @@
 
 import fish.payara.cdi.auth.roles.CallerAccessException;
 import fish.payara.cdi.auth.roles.RolesPermitted;
+import java.lang.reflect.Parameter;
+import javax.el.ELProcessor;
+import javax.inject.Named;
+import javax.security.enterprise.AuthenticationStatus;
+import static javax.security.enterprise.AuthenticationStatus.NOT_DONE;
```
Comment (dmatej) [inline appserver/payara-appserver-modules/cdi-auth-roles/src/main/java/fish/payara/appserver/cdi/auth/roles/RolesPermittedInterceptor.java]: Again that chaotic import ordering ...

File: appserver/payara-appserver-modules/cdi-auth-roles/src/main/java/fish/payara/appserver/cdi/auth/roles/RolesPermittedInterceptor.java [hunk 2]
```diff
@@ -193,13 +230,58 @@ private RolesPermitted getRolesPermitted(InvocationContext invocationContext) {
...
+                && roles.length > 0 && !isAuthenticated()) {
+            AuthenticationStatus status = securityContext.authenticate(request, response, withParams());
+
+            // Authentication was not done at all (i.e. no credentials present) or
+            // authentication failed (i.e. wrong credentials, credentials expired, etc)
+            if (status == NOT_DONE || status == SEND_FAILURE) {
+                throw new NotAuthorizedException("Authentication resulted in " + status);
+            }
+
+            if (status == SUCCESS && !isAuthenticated()) { // compensate for possible Soteria bug, need to investigate
```
Comment (dmatej) [inline appserver/payara-appserver-modules/cdi-auth-roles/src/main/java/fish/payara/appserver/cdi/auth/roles/RolesPermittedInterceptor.java]: Better move the comment before the if

File: appserver/payara-appserver-modules/cdi-auth-roles/src/main/java/fish/payara/appserver/cdi/auth/roles/RolesPermittedInterceptor.java (hunk 1 above)
Comment (jGauravGupta) [reply appserver/payara-appserver-modules/cdi-auth-roles/src/main/java/fish/payara/appserver/cdi/auth/roles/RolesPermittedInterceptor.java]: That's automatically ordered by IDE in alphabetical order of package name.

**Comment by jGauravGupta** [comment]
jenkins test please

**Pull request merged**